exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go

```diff
@@ -19,65 +19,381 @@ package controllers
+
+		return ctrl.Result{}, nil
+	case snapshotrestorev1.ETCDSnapshotRestorePhaseStarted:
+		etcdSnapshotRestore.Status.Phase = snapshotrestorev1.ETCDSnapshotRestorePhaseShutdown
+
+		return ctrl.Result{}, nil
+	case snapshotrestorev1.ETCDSnapshotRestorePhaseShutdown:
+		// Stop RKE2 on all the machines.
+		return r.stopRKE2OnAllMachines(ctx, scope, etcdSnapshotRestore)
+	case snapshotrestorev1.ETCDSnapshotRestorePhaseRunning:
+		// Restore the etcd snapshot on the init machine.
+		return r.restoreSnaphotOnInitMachine(ctx, scope, etcdSnapshotRestore)
+	case snapshotrestorev1.ETCDSnapshotRestorePhaseAgentRestart:
+		// Start RKE2 on all the machines.
+		return r.startRKE2OnAllMachines(ctx, scope, etcdSnapshotRestore)
+	case snapshotrestorev1.ETCDSnapshotRestoreUnpauseCluster:
+		// Unpause the cluster to reconcile machine agent conditions
+		scope.cluster.Spec.Paused = false
+		etcdSnapshotRestore.Status.Phase = snapshotrestorev1.ETCDSnapshotRestorePhaseJoinAgents
+
+		return ctrl.Result{}, nil
+	case snapshotrestorev1.ETCDSnapshotRestorePhaseJoinAgents:
+		return r.waitForMachinesToJoin(ctx, scope, etcdSnapshotRestore)
+	case snapshotrestorev1.ETCDSnapshotRestorePhaseFinished, snapshotrestorev1.ETCDSnapshotRestorePhaseFailed:
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ETCDSnapshotRestoreReconciler) reconcileDelete(_ context.Context, _ *snapshotrestorev1.ETCDSnapshotRestore) (ctrl.Result, error) {
+	return ctrl.Result{}, nil
+}
+
+func initScope(ctx context.Context, c client.Client, etcdSnapshotRestore *snapshotrestorev1.ETCDSnapshotRestore) (*scope, error) {
+	// Get the cluster object.
+	cluster := &clusterv1.Cluster{}
+
+	if err := c.Get(ctx, client.ObjectKey{Namespace: etcdSnapshotRestore.Namespace, Name: etcdSnapshotRestore.Spec.ClusterName}, cluster); err != nil {
+		return nil, fmt.Errorf("failed to get cluster: %w", err)
+	}
+
+	machines, err := collections.GetFilteredMachinesForCluster(ctx, c, cluster)
+	if err != nil {
+		return nil, fmt.Errorf("failed to collect machines for cluster: %w", err)
+	}
+
+	// Get etcd machine backup object.
+	etcdMachineSnapshot := &snapshotrestorev1.ETCDMachineSnapshot{}
+	if err := c.Get(ctx, client.ObjectKey{
+		Namespace: etcdSnapshotRestore.Namespace,
+		Name:      etcdSnapshotRestore.Spec.ETCDMachineSnapshotName,
+	}, etcdMachineSnapshot); err != nil {
+		return nil, fmt.Errorf("failed to get etcd machine backup: %w", err)
+	}
+
+	return &scope{
+		cluster:             cluster,
+		machines:            machines.Filter(collections.ControlPlaneMachines(cluster.Name)),
+		etcdMachineSnapshot: etcdMachineSnapshot,
+	}, nil
+}
+
+func (r *ETCDSnapshotRestoreReconciler) stopRKE2OnAllMachines(ctx context.Context, scope *scope, etcdSnapshotRestore *snapshotrestorev1.ETCDSnapshotRestore) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	results := []Output{}
+	for _, machine := range scope.machines {
```

Comment:
based on recent discussions, we can stop RKE2 service only on control planes

Reply:
I think currently the set of machines already limited to CP. Only agent installation is performed on worker nodes.